### DISCUSSION TITLE
Why does the random seed need many ones?

### DISCUSSION THREAD

**Discussion by fepegar** [Project-MONAI/MONAI#2149] [Q&A]
The docs for [`monai.utils.misc.set_determinism`](https://docs.monai.io/en/latest/utils.html#monai.utils.misc.set_determinism) say:

> It is recommended to set a large seed, i.e. a number that has a good balance of 0 and 1 bits. Avoid having many 0 bits in the seed.

Two things confuse me here:
1. Why is a good balance between 0s and 1s better? Why is having many 0s bad? I haven't found any sources to support these claims. Moreover, `0` is the most common random seed ([source](https://twitter.com/jakevdp/status/1247742792861757441)) and some important libraries use it happily in their examples.
2. A large seed doesn't imply a good balance, nor a lot of ones. The current default is `np.iinfo(np.uint32).max`, which is `4294967295` or `11111111111111111111111111111111`. Another large number is `2147483648` or `10000000000000000000000000000000`. The default kwarg is not consistent with the docstring.

**Comment by Nic-Ma** [comment]
Hi @wyli ,

Could you please help share some comments about this question？

Thanks in advance.

**Comment by wyli** [comment]
Thanks, for item 1, the docstring is probably from pytorch https://pytorch.org/docs/stable/generated/torch.Generator.html#torch.Generator.manual_seed
and this "[Good Practice in (Pseudo) Random Number Generation for
Bioinformatics Applications](http://www0.cs.ucl.ac.uk/staff/d.jones/GoodPracticeRNG.pdf)" suggests: 
```
This is important if your seed
values have very low entropy (or more simply they have too many runs of 0 or 1 bits in their binary
representation) – some otherwise good RNGs (especially shift-register-based generators such as MT)
are not very random from their initial state when certain seed values are used. 
```

I think item 2 is a valid issue and I don't know the reason of the current default... following the good practice doc, we might want to 'warm up' the RNGs. cc @ericspod @rijobro

**Reply by rijobro** [reply]
Interesting read, thanks @wyli! Looks like warming up might be the simplest solution, we could an extra argument to the `set_determinism`/`set_random_state` functions. The only tricky thing I see initially is that we'd have to warm up all of our random libraries equally (python, np and torch).

**Reply by fepegar** [reply]
Thanks, Wenqi. I think PyTorch docs on this are not good, then. Given that [no one seems to care much](https://github.com/pytorch/pytorch/search?q=torch.manual_seed) about this (even [Soumith Chintala](https://github.com/soumith/traffic-sign-detection-homework/blob/8c223326d7f2bf2c52a97f4e41bd1a0e9bf59bae/main.py#L22-L28)), it might be an unnecessary worry. Shall we ask the PyTorch devs about this issue?
